File: android/src/main/java/com/sunmi/v2/printer/SunmiV2PrinterModule.java
package com.sunmi.v2.printer;

import android.content.BroadcastReceiver;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;

import java.util.Map;

import android.os.RemoteException;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.util.Base64;
import android.graphics.Bitmap;
import android.util.Log;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.IntentFilter;


import com.sunmi.peripheral.printer.InnerPrinterCallback;
import com.sunmi.peripheral.printer.InnerPrinterException;
import com.sunmi.peripheral.printer.InnerPrinterManager;
import com.sunmi.peripheral.printer.InnerResultCallback;
import com.sunmi.peripheral.printer.SunmiPrinterService;

class MyResultCallback extends InnerResultCallback {
    private com.facebook.react.bridge.Promise myPromise;
    private String myTag = "MyResultCallback";

    @Override
    public void onRunResult(boolean isSuccess) throws RemoteException {
        Log.i(myTag, "onRunResult: " + isSuccess);
        if (myPromise != null) {
            myPromise.resolve(isSuccess);
            myPromise = null;
        }
    }

    @Override
    public void onReturnString(String result) throws RemoteException {
        Log.i(myTag, "onReturnString: " + result);
        if (myPromise != null) {
            myPromise.resolve(result);
            myPromise = null;
        }
    }

    @Override
    public void onRaiseException(int code, String msg) throws RemoteException {
        Log.i(myTag, "onRaiseException: " + msg);
        if (myPromise != null) {
            myPromise.reject(String.valueOf(code), msg);
            myPromise = null;
        }
    }

    @Override
    public void onPrintResult(int code, String msg) throws RemoteException {
        Log.i(myTag, "onPrintResult: " + msg);
        if (myPromise != null) {
            myPromise.resolve(msg);
            myPromise = null;
        }
    }

    public void setPromise(Promise promise) {
        this.myPromise = promise;
    }
}

public class SunmiV2PrinterModule extends ReactContextBaseJavaModule {
    public static ReactApplicationContext reactApplicationContext = null;
    private BitmapUtils bitMapUtils;
    private PrinterReceiver receiver = new PrinterReceiver();

    public static String noPrinter = "NO_PRINTER_INIT";
    public static int NoSunmiPrinter = 0x00000000;
    public static int CheckSunmiPrinter = 0x00000001;
    public static int FoundSunmiPrinter = 0x00000002;
    public static int LostSunmiPrinter = 0x00000003;

    private MyResultCallback innerResultCallback = new MyResultCallback();

    /**
     *  sunmiPrinter means checking the printer connection status
     */
    public int sunmiPrinter = CheckSunmiPrinter;

    private static SunmiPrintHelper helper = new SunmiPrintHelper();

    public static SunmiPrintHelper getInstance() {
        return helper;
    }

    private ServiceConnection connService = new ServiceConnection() {
        @Override
        public void onServiceDisconnected(ComponentName name) {
            Log.i(TAG, "Service disconnected: " + name);
        }

        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            Log.i(TAG, "Service connected: " + name);

        }
    };

    private static final String TAG = "SunmiV2PrinterModule";

    public SunmiV2PrinterModule(ReactApplicationContext reactContext) {
        super(reactContext);
        reactApplicationContext = reactContext;
        bitMapUtils = new BitmapUtils(reactContext);
        IntentFilter mFilter = new IntentFilter();
        getReactApplicationContext().registerReceiver(receiver, mFilter);
    }

    @Override
    public String getName() {
        return "SunmiV2Printer";
    }

    @ReactMethod
    public void initBind(final Promise promise) {
        try {
            SunmiPrintHelper.getInstance().initSunmiPrinterService(reactApplicationContext);
            promise.resolve(null);
        } catch (Exception e) {
            Log.i(TAG, "ERROR: " + e.getMessage());
            promise.reject("0", e.getMessage());
        }
    }

    @ReactMethod
    public void getPrinterDidBind(final Promise p) {
        try {
            p.resolve(SunmiPrintHelper.getInstance().printerDidBind());
        } catch (Exception e) {
            Log.i(TAG, "ERROR: " + e.getMessage());
            p.reject("0", e.getMessage());
        }
    }

    @ReactMethod
    public void initPrinter(final Promise promise) {
        try {
            SunmiPrintHelper.getInstance().initPrinter();
            promise.resolve(null);
        } catch (Exception e) {
            Log.i(TAG, "ERROR: " + e.getMessage());
            promise.reject("0", e.getMessage());
        }
    }

    @ReactMethod
    public void getPrinterServiceDidInit(final Promise p) {
        try {
            p.resolve(SunmiPrintHelper.getInstance().printerServiceDidInit());
        } catch (Exception e) {
            Log.i(TAG, "ERROR: " + e.getMessage());
            p.reject("0", e.getMessage());
        }
    }

    @ReactMethod
    public void getPrinterStatus(final Promise p) {
         try {
             p.resolve(SunmiPrintHelper.getInstance().getPrinterStatus());
         } catch (Exception e) {
             Log.i(TAG, "ERROR: " + e.getMessage());
             p.reject("0", e.getMessage());
         }
    }

    @ReactMethod
    public void getPrinterVersion(final Promise p) {
         try {
             p.resolve(SunmiPrintHelper.getInstance().getPrinterVersion());
         } catch (Exception e) {
             Log.i(TAG, "ERROR: " + e.getMessage());
             p.reject("0", e.getMessage());
         }
    }

    @ReactMethod
    public void getPrinterPaperSize(final Promise p) {
        try {
            p.resolve(SunmiPrintHelper.getInstance().getPrinterPaper());
        } catch (Exception e) {
            Log.i(TAG, "ERROR: " + e.getMessage());
            p.reject("0", e.getMessage());
        }
    }

    @ReactMethod
    public void getPrinterSpecifications(final Promise p) {
        innerResultCallback.setPromise(p);
        try {
            SunmiPrintHelper.getInstance().getPrinterHead(innerResultCallback);
        } catch (Exception e) {
            Log.i(TAG, "ERROR: " + e.getMessage());
            p.reject("0", e.getMessage());
        }
    }

    @ReactMethod
    public void getPrinterDensity(final Promise p) {
        try {
            p.resolve(SunmiPrintHelper.getInstance().getPrinterDensity());
        } catch (Exception e) {
            Log.i(TAG, "ERROR: " + e.getMessage());
            p.reject("0", e.getMessage());
        }
    }

    @ReactMethod
    public void getServiceVersion(final Promise p) {
        try {
            p.resolve(SunmiPrintHelper.getInstance().getServiceVersion());
        } catch (Exception e) {
            Log.i(TAG, "ERROR: " + e.getMessage());
            p.reject("0", e.getMessage());
        }
    }

    @ReactMethod
    public void getPrinterDistance(final Promise p) {
        innerResultCallback.setPromise(p);
        try {
            SunmiPrintHelper.getInstance().getPrinterDistance(innerResultCallback);
        } catch (Exception e) {
            Log.i(TAG, "ERROR: " + e.getMessage());
            p.reject("0", e.getMessage());
        }
    }

    @ReactMethod
    public void showPrinterStatusToast(final Promise p) {
        try {
            SunmiPrintHelper.getInstance().showPrinterStatus(reactApplicationContext);
            p.resolve(null);
        } catch (Exception e) {
            Log.i(TAG, "ERROR: " + e.getMessage());
            p.reject("0", e.getMessage());
        }
    }

    @ReactMethod
    public void sendRAWData(String base64EncriptedData, final Promise promise) {
        try {
            final byte[] data = Base64.decode(base64EncriptedData, Base64.DEFAULT);
            SunmiPrintHelper.getInstance().sendRawData(data);
            promise.resolve(null);

        } catch (Exception e) {
            Log.i(TAG, "ERROR: " + e.getMessage());
            promise.reject("0", e.getMessage());
        }
    }

    @ReactMethod
    public void setAlignment(int alignment, final Promise promise) {
        try {
            SunmiPrintHelper.getInstance().setAlign(alignment);
            promise.resolve(null);
        } catch (Exception e) {
            Log.i(TAG, "ERROR: " + e.getMessage());
            promise.reject("0", e.getMessage());
        }
    }

    @ReactMethod
    public void printBitmap(String data, int width, int height, final Promise promise) {

        try {
            byte[] decoded = Base64.decode(data, Base64.DEFAULT);
            final Bitmap bitMap = bitMapUtils.decodeBitmap(decoded, width, height);

            SunmiPrintHelper.getInstance().printBitmap(bitMap);

            promise.resolve(null);
        } catch (Exception e) {
            e.printStackTrace();
            promise.reject("0", e.getMessage());
        }
    }

    @ReactMethod
    public void printText(String text, final Promise promise) {
        try {
            SunmiPrintHelper.getInstance().printText(text);
            promise.resolve(null);
        } catch (Exception e) {
            e.printStackTrace();
            promise.reject("0", e.getMessage());
        }
    }

    @ReactMethod
    public void clearBuffer(final Promise promise) {
        try {
            SunmiPrintHelper.getInstance().deInitSunmiPrinterService(reactApplicationContext);
            promise.resolve(null);
        } catch (Exception e) {
            e.printStackTrace();
            promise.reject("0", e.getMessage());
        }
    }

    @ReactMethod
    public void openCashDrawer(final Promise promise) {
        try {
            SunmiPrintHelper.getInstance().openCashBox();
            promise.resolve(null);

        } catch (Exception e) {
            e.printStackTrace();
            promise.reject("0", e.getMessage());

        }
    }

    @ReactMethod
    public void cutPaper(final Promise promise) {
        try {
            SunmiPrintHelper.getInstance().cutpaper();
            promise.resolve(null);
        } catch (Exception e) {
            e.printStackTrace();
            promise.reject("0", e.getMessage());
        }
    }
}

File: android/src/main/java/expo/modules/sunmiv2printer/SunmiV2PrinterModule.kt
package expo.modules.sunmiv2printer

import SunmiResultCallback
import android.graphics.Bitmap
import android.util.Base64
import android.util.Log
import com.facebook.react.bridge.ReactApplicationContext
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class SunmiV2PrinterModule : Module() {
    public companion object {
        @kotlin.jvm.JvmField
        public var reactApplicationContext: ReactApplicationContext? = null
    }

    private val helper = SunmiPrintHelper();
    private val TAG = "SunmiV2PrinterModule"
    private val innerResultCallback = SunmiResultCallback()
    // private var bitMapUtils: BitmapUtils? = null

    // Each module class must implement the definition function. The definition consists of components
    // that describes the module's functionality and behavior.
    // See https://docs.expo.dev/modules/module-api for more details about available components.
    override fun definition() = ModuleDefinition {
        // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
        // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
        // The module will be accessible from `requireNativeModule('SunmiV2PrinterModule')` in JavaScript.
        Name(TAG)

        AsyncFunction("initBind") { promise: Promise ->
            Log.i(TAG, "initBind");
            try {
                if (appContext != null) {
                    promise.resolve(helper.initSunmiPrinterService(appContext.reactContext))
                } else {
                    promise.reject("no context")
                }
            } catch (e: Exception) {
                Log.i(TAG, "initBind did fail:" + e.message);
                promise.reject(e.message ?: "Error")
            }
        }

        AsyncFunction("getPrinterDidBind") { promise: Promise ->
            Log.i(TAG, "getPrinterDidBind");
            try {
                promise.resolve(helper.printerDidBind())
            } catch (e: Exception) {
                Log.i(TAG, "getPrinterDidBind did fail:" + e.message);
                promise.reject(e.message ?: "Error")
            }
        }

        AsyncFunction("initPrinter") { promise: Promise ->
            Log.i(TAG, "initPrinter");
            try {
                promise.resolve(helper.initPrinter())
            } catch (e: Exception) {
                Log.i(TAG, "initPrinter did fail:" + e.message);
                promise.reject(e.message ?: "initPrinterError")
            }
        }

        AsyncFunction("getPrinterServiceDidInit") { promise: Promise ->
            Log.i(TAG, "getPrinterServiceDidInit");
            try {
                promise.resolve(helper.printerServiceDidInit())
            } catch (e: Exception) {
                Log.i(TAG, "getPrinterServiceDidInit did fail:" + e.message);
                promise.reject(e.message ?: "Error")
            }
        }


        AsyncFunction("getPrinterVersion") { promise: Promise ->
            Log.i(TAG, "getPrinterVersion");
            try {
                promise.resolve(helper.printerVersion)
            } catch (e: Exception) {
                Log.i(TAG, "getPrinterVersion did fail:" + e.message);
                promise.reject(e.message ?: "Error")
            }
        }


        AsyncFunction("getPrinterPaperSize") { promise: Promise ->
            Log.i(TAG, "getPrinterPaperSize");
            try {
                promise.resolve(helper.printerPaper)
            } catch (e: Exception) {
                Log.i(TAG, "getPrinterVersion did fail:" + e.message);
                promise.reject(e.message ?: "Error")
            }
        }

        AsyncFunction("getPrinterSpecifications") { promise: Promise ->
            Log.i(TAG, "getPrinterSpecifications");
            try {
                promise.resolve(helper.getPrinterHead(innerResultCallback))
            } catch (e: Exception) {
                Log.i(TAG, "getPrinterSpecifications did fail:" + e.message);
                promise.reject(e.message ?: "Error")
            }
        }

        AsyncFunction("getPrinterDensity") { promise: Promise ->
            Log.i(TAG, "getPrinterDensity");
            try {
                promise.resolve(helper.printerDensity)
            } catch (e: Exception) {
                Log.i(TAG, "getPrinterSpecifications did fail:" + e.message);
                promise.reject(e.message ?: "Error")
            }
        }

        AsyncFunction("getServiceVersion") { promise: Promise ->
            Log.i(TAG, "getServiceVersion");
            try {
                promise.resolve(helper.serviceVersion)
            } catch (e: Exception) {
                Log.i(TAG, "getServiceVersion did fail:" + e.message);
                promise.reject(e.message ?: "Error")
            }
        }

        AsyncFunction("getPrinterDistance") { promise: Promise ->
            Log.i(TAG, "getPrinterDistance");
            try {
                promise.resolve(helper.getPrinterDistance(innerResultCallback))
            } catch (e: Exception) {
                Log.i(TAG, "getPrinterDistance did fail:" + e.message);
                promise.reject(e.message ?: "Error")
            }
        }

        AsyncFunction("showPrinterStatusToast") { promise: Promise ->
            Log.i(TAG, "showPrinterStatusToast");
            try {
                promise.resolve(helper.showPrinterStatus(appContext.reactContext))
            } catch (e: Exception) {
                Log.i(TAG, "showPrinterStatusToast did fail:" + e.message);
                promise.reject(e.message ?: "Error")
            }
        }

        AsyncFunction("sendRAWData") { base64EncriptedData: String, promise: Promise ->
            Log.i(TAG, "sendRAWData");
            try {
                val data: ByteArray = Base64.decode(base64EncriptedData, Base64.DEFAULT)
                promise.resolve(helper.sendRawData(data))
            } catch (e: Exception) {
                Log.i(TAG, "sendRAWData did fail:" + e.message);
                promise.reject(e.message ?: "Error")
            }
        }

        AsyncFunction("setAlignment") { alignment: Int, promise: Promise ->
            Log.i(TAG, "sendRAWData");
            try {
                promise.resolve(helper.setAlign(alignment))
            } catch (e: Exception) {
                Log.i(TAG, "setAlignment did fail:" + e.message);
                promise.reject(e.message ?: "Error")
            }
        }

        AsyncFunction("printBitmap") { data: String, width: Int, height: Int, promise: Promise ->
            Log.i(TAG, "printBitmap");
            try {
                val decoded: ByteArray = Base64.decode(data, Base64.DEFAULT)
                if (appContext != null) {
                    val bitMap: Bitmap =
                        BitmapUtils(appContext.reactContext)!!.decodeBitmap(decoded, width, height)
                    promise.resolve(helper.printBitmap(bitMap))
                } else {
                    promise.reject("no context")
                }
            } catch (e: Exception) {
                Log.i(TAG, "printBitmap did fail:" + e.message);
                promise.reject(e.message ?: "Error")
            }
        }

        AsyncFunction("printText") { text: String, promise: Promise ->
            Log.i(TAG, "printText");
            try {
                promise.resolve(helper.printText(text))
            } catch (e: Exception) {
                Log.i(TAG, "printText did fail:" + e.message);
                promise.reject(e.message ?: "Error")
            }
        }

        AsyncFunction("clearBuffer") { promise: Promise ->
            Log.i(TAG, "clearBuffer");
            try {
                promise.resolve(helper.deInitSunmiPrinterService(appContext.reactContext))
            } catch (e: Exception) {
                Log.i(TAG, "clearBuffer did fail:" + e.message);
                promise.reject(e.message ?: "Error")
            }
        }

        AsyncFunction("openCashDrawer") { promise: Promise ->
            Log.i(TAG, "openCashDrawer");
            try {
                promise.resolve(helper.openCashBox())
            } catch (e: Exception) {
                Log.i(TAG, "openCashDrawer did fail:" + e.message);
                promise.reject(e.message ?: "Error")
            }
        }

        AsyncFunction("cutPaper") { promise: Promise ->
            Log.i(TAG, "cutPaper");
            try {
                promise.resolve(helper.cutpaper())
            } catch (e: Exception) {
                Log.i(TAG, "cutPaper did fail:" + e.message);
                promise.reject(e.message ?: "Error")
            }
        }
    }
}
